### ISSUE TITLE
Multicore & Test Data Queries?

### ISSUE THREAD

**Issue by stnava** [jlmelville/rcpphnsw#7]
Hi - thanks for this package - very useful.  Just have two questions:

(1) Why no multithreading support?  

(2) Is it possible to add something like:

FNN::get.knnx( t( x ), t( y ), k=k )

i.e. the ability to query against test data ( y )?

If not, why not?

Thanks for your insights on these questions.

Best,

BA

**Comment by LTLA** [comment]
For **2**: doesn't `hnsw_search` do the job for you? Alternatively, another implementation is provided by `queryKNN` in [_BiocSingular_](https://bioconductor.org/packages/devel/bioc/html/BiocSingular.html).

For **1**, I don't know about why multithreading was removed in a992c834a5726b91e582fce60bbcf337fe59e9de, I would guess that it was more trouble than it was worth. I'll just point out that _BiocSingular_ supports a more coarse-grained parallelization via _BiocParallel_.

**Comment by stnava** [comment]
re: 2 - yes => thanks for pointing that out.  here is an example:

```
irismx <- as.matrix(iris[1:100, -5])
irismy <- as.matrix(iris[101:150, -5])
ann <- hnsw_build(irismx)
iris_nn <- hnsw_search(irismy, ann, k = 5)
```

re: both 1 and 2, my knn searches are very large so I am looking for the fastest possible memory efficient methods I can find.

These answers are sufficient for me to close.

**Comment by jlmelville** [comment]
I see I have been beaten to the punch with responding.

I see that the man page text for `hnsw_search` is a bit misleading and only talks about searching against the same data you used to build the index. I must change that.

For multi-threading, I took it out because I couldn't get it to work: it persistently caused segfaults. This is almost certainly because of a mistake I made, but I ran out of enthusiasm to track it down before submitting to CRAN. I will look at bringing it back for a subsequent release. But this will only affect searching the index, building the index will probably always be single-threaded.
